x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -233,16 +211,82 @@ public void finalizeSnapshot(
+            assert computedHash.indexOf('.') == -1;
+            return seed + "." + uuid + "." + computedHash;
+        } catch (Exception e) {
+            throw new RepositoryVerificationException(metadata.name(), "Exception computing password hash for repository verification", e);
+        }
+    }
+
+    @Override
+    public void endVerification(String packedSeed) {
+        // verification bypasses the encrypted blobstore because it works easier without a password generation
+        String seed = packedSeed.substring(0, packedSeed.indexOf('.'));
+        super.endVerification(seed);
+    }
+
+    @Override
+    public void verify(String packedSeed, DiscoveryNode localNode) {
+        // verification bypasses the encrypted blobstore because it works easier without a password generation
+        String seed = packedSeed.substring(0, packedSeed.indexOf('.'));
+        String uuid = packedSeed.substring(seed.length() + 1, packedSeed.lastIndexOf('.'));
+        String hash = packedSeed.substring(packedSeed.lastIndexOf('.') + 1);
+        try {
+            String computedHash = AESKeyUtils.computeId(AESKeyUtils.generatePasswordBasedKey(repositoryPassword, uuid));
+            if (false == hash.equals(computedHash)) {
+                throw new ElasticsearchException("Password mismatch on node {}", localNode);
+            }
+        } catch (Exception e) {
+            throw new RepositoryVerificationException(metadata.name(), "Error verifying password hash", e);
+        }
+        super.verify(seed, localNode);
+    }
+
+    @Override
+    public BlobContainer testBlobContainer(String seed) {
+        // bypass encryption for test blobs because the lifecycle of password generation is a bit complex
+        // when accounting for optional repository verification and it is not very useful
+        EncryptedBlobStore encryptedBlobStore = (EncryptedBlobStore) blobStore();
+        return encryptedBlobStore.delegatedBlobStore.blobContainer(basePath().add(testBlobPrefix(seed)));
+    }
+
+    @Override
+    public void snapshotShard(SnapshotShardContext context) {
+        if (false == licenseStateSupplier.get().isAllowed(XPackLicenseState.Feature.ENCRYPTED_SNAPSHOT)) {
+            context.onFailure(LicenseUtils.newComplianceException("encrypted snapshots"));
             return;
         }
         super.snapshotShard(context);
     }
 
+    @Override
+    public void executeConsistentStateUpdate(
+        BiConsumer<RepositoryData, ActionListener<ClusterStateUpdateTask>> createUpdateTaskAsync,
+        String source,
+        Consumer<Exception> onFailure
+    ) {
+        super.executeConsistentStateUpdate((repositoryData, createUpdateTaskListener) -> {
+            if (repositoryData.equals(RepositoryData.EMPTY) && false == isReadOnly()) {
+                threadPool().generic().execute(ActionRunnable.wrap(createUpdateTaskListener, (l) -> {
```

Comment:
As discussed, this should be moved on to the specially devised pbkdf2 threadpool, the one we're currently only using for API Keys.